frost-core/src/frost.rs

```diff
@@ -148,34 +148,67 @@ where
     }
 }
 
-/// Generates the lagrange coefficient for the i'th participant.
-#[cfg_attr(feature = "internals", visibility::make(pub))]
-fn derive_interpolating_value<C: Ciphersuite>(
-    signer_id: &Identifier<C>,
-    signing_package: &SigningPackage<C>,
+/// Generates a lagrange coefficient.
+///
+/// The Lagrange polynomial for a set of points (x_j, y_j) for 0 <= j <= k
+/// is ∑_{i=0}^k y_i.ℓ_i(x), where ℓ_i(x) is the Lagrange basis polynomial:
+///
+/// ℓ_i(x) = ∏_{0≤j≤k; j≠i} (x - x_j) / (x_i - x_j).
+///
+/// This computes ℓ_j(x) for the set of points `xs` and for the j corresponding
+/// to the given xj.
+///
+/// If `x` is None, it uses 0 for it (since Identifiers can't be 0)
+pub fn compute_lagrange_coefficient<C: Ciphersuite>(
+    xs: &BTreeSet<Identifier<C>>,
+    x: Option<Identifier<C>>,
+    xi: Identifier<C>,
 ) -> Result<Scalar<C>, Error<C>> {
-    let zero = <<C::Group as Group>::Field>::zero();
-
+    if xs.is_empty() {
+        return Err(Error::IncorrectNumberOfIdentifiers);
+    }
     let mut num = <<C::Group as Group>::Field>::one();
     let mut den = <<C::Group as Group>::Field>::one();
 
-    for commitment_identifier in signing_package.signing_commitments().keys() {
-        if *commitment_identifier == *signer_id {
+    let mut xi_found = false;
+    for xj in xs.iter() {
+        if xi == *xj {
+            xi_found = true;
             continue;
         }
 
-        num *= *commitment_identifier;
-
-        den *= *commitment_identifier - *signer_id;
+        if let Some(x) = x {
+            num *= x - *xj;
+            den *= xi - *xj;
+        } else {
+            // Both signs inverted just to avoid requiring Neg (-*xj)
+            num *= *xj;
+            den *= *xj - xi;
+        }
     }
-
-    if den == zero {
-        return Err(Error::DuplicatedShares);
+    if !xi_found {
```

Comment:
Should we return this immediately after the `xi_found` loop above?

Reply:
I think we can't, we are determining if x_i is inside the set or not. We'll only be sure after we iterate the entire set.

I just realized this is a microoptimization though, we can just check if `!x_set.contains(&x_i)` before the loop if you prefer